src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,94 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-
-            // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            if (x is not null && y is not null)
             {
-                x = xParts[i];
-                y = yParts[i];
+                if (x.Contains(ConfigurationPath.KeyDelimiter) && y.Contains(ConfigurationPath.KeyDelimiter))
+                {
+                    int xIndex = 0;
+                    int yIndex = 0;
+
+                    // Compare each part until we get two parts that are not equal
+                    while (xIndex < x.Length && yIndex < y.Length)
+                    {
+                        int nextXIndex = x.IndexOf(ConfigurationPath.KeyDelimiter, xIndex);
+                        int nextYIndex = y.IndexOf(ConfigurationPath.KeyDelimiter, yIndex);
+
+                        ReadOnlySpan<char> xSpan;
+                        if (nextXIndex >= 0)
+                        {
+                            xSpan = x.AsSpan().Slice(xIndex, nextXIndex - xIndex);
+                            xIndex = nextXIndex + 1;
+                        }
+                        else
+                        {
+                            xSpan = x.AsSpan().Slice(xIndex, x.Length - xIndex);
+                            xIndex = x.Length;
+                        }
 
-                int value1 = 0;
-                int value2 = 0;
+                        ReadOnlySpan<char> ySpan;
+                        if (nextYIndex >= 0)
+                        {
+                            ySpan = y.AsSpan().Slice(yIndex, nextYIndex - yIndex);
+                            yIndex = nextYIndex + 1;
+                        }
+                        else
+                        {
+                            ySpan = y.AsSpan().Slice(yIndex, y.Length - yIndex);
+                            yIndex = y.Length;
+                        }
+
+                        int compareResult = Compare(xSpan, ySpan);
+                        if (compareResult != 0)
+                        {
+                            return compareResult;
+                        }
+                    }
+
+                    string[] xParts = x.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries);
+                    string[] yParts = y.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
why keeping these lines?

Reply:
It seems like we could just keep a count of x parts and y parts above, and then compare them here instead of splitting again.